util/chunk/chunk.go

```diff
@@ -112,12 +156,44 @@ func (c *Chunk) SetNumVirtualRows(numVirtualRows int) {
 // Reset resets the chunk, so the memory it allocated can be reused.
 // Make sure all the data in the chunk is not used anymore before you reuse this chunk.
 func (c *Chunk) Reset() {
-	for _, c := range c.columns {
-		c.reset()
+	for _, col := range c.columns {
+		col.reset()
 	}
 	c.numVirtualRows = 0
 }
 
+// GrowAndReset resets the chunk and grow capacity if needed.
+// TODO: this method will be used in continue PR.
+func (c *Chunk) GrowAndReset(maxChunkSize int) {
+	if c.columns == nil {
+		return
+	}
+	newCap := calculateCapacity(c, maxChunkSize)
+	if newCap <= c.capacity {
+		c.Reset()
+		return
+	}
+	c.capacity = newCap
+	c.columns = renewColumns(c.columns, newCap)
+	c.numVirtualRows = 0
+}
+
+// calculateCapacity suitable capacity in next execution.
+//      using capacity limit if full but capacity already meet maxChunkSize
+//      using oldcap x 2 if chk is full but less than maxChunkSize
+//      keep oldcap if chk isn't full
+func calculateCapacity(c *Chunk, maxChunkSize int) int {
```

Comment:
how about s/calculateCapacity/reCalcCapacity/ and changing the comment to:
```go
// reCalcCapacity calculates the capacity for another Chunk based the current
// Chunk. The new capacity is doubled only when the current Chunk is full.
```

Reply:
Based on Jason's suggestion:

// reCalcCapacity calculates the capacity for another Chunk based **on** the current
// Chunk. The new capacity is doubled only when the current Chunk is full.